File: src/configs.py
import json

from typing import Tuple, Dict, Union
from dataclasses import dataclass, asdict


@dataclass
class Description:
    term: str = "number of days to split the seed and buy the stock"
    margin: str = "margin to sell the stock"
    bullish_rsi: str = "rsi threshold to buy the stock"
    bullish_u50: str = (
        "rate of market days under 50 moving average to determine burst buy and "
    )
    burst_scale: str = "scale of burst buy when the market fluctuates"
    burst_vol: str = "volatility threshold to determine burst buy"
    sell_base: str = "base sell rate when all seed is exhausted"
    sell_limit: str = "sell limit when all seed is exhausted"
    sahm_threshold: str = "sahm threshold to exclude when sliding window test"

    split: str = "number to split the seed"
    buy_rsi: str = "rsi threshold to buy the stock"
    sell_rsi: str = "rsi threshold to sell the stock"


@dataclass
class Bounds:
    term: Tuple[int] = (40, 40)
    margin: Tuple[float] = (0.05, 0.15)
    bullish_rsi: Tuple[int] = (60, 100)
    bullish_u50: Tuple[float] = (0.3, 0.8)
    burst_scale: Tuple[float] = (0.0, 3.0)
    burst_vol: Tuple[int] = (25, 50)
    sell_base: Tuple[float] = (0, 0.5)
    sell_limit: Tuple[float] = (0.5, 1.0)
    sahm_threshold: Tuple[float] = (1.0, 1.0)
    split: Tuple[int] = (10, 20)
    buy_rsi: Tuple[int] = (20, 50)
    sell_rsi: Tuple[int] = (70, 100)


@dataclass
class Precisions:
    term: int = 1
    margin: float = 0.01
    bullish_rsi: int = 5
    bullish_u50: float = 0.1
    burst_scale: float = 0.5
    burst_vol: int = 5
    sell_base: float = 0.1
    sell_limit: float = 0.1
    sahm_threshold: float = 0.5
    split: int = 5
    buy_rsi: int = 5
    sell_rsi: int = 5


@dataclass
class Config:
    @classmethod
    def _from(cls, map: Dict[str, Union[float, int]]) -> "Config":
        inst = cls()

        for field, value in map.items():
            if hasattr(inst, field):
                setattr(inst, field, value)

        return inst

    def __hash__(self):
        h = 0
        for k, v in asdict(self).items():
            h = hash((h, k, v))

        return h

    def __str__(self):
        l = []
        for k, v in asdict(self).items():
            vs = f"{v:.2f}" if isinstance(v, float) else f"{v}"
            l.append(f"{k}: {vs}")

        return ", ".join(l)


@dataclass
class V1Config(Config):
    term: int = 40
    margin: float = 0.1  # [0.05:0.15:0.01] : 11
    bullish_rsi: int = 80  # [40:100:5] : 13
    bullish_u50: float = 0.5
    burst_scale: float = 0.0
    burst_vol: int = 30
    sell_base: float = 0
    sell_limit: float = 1
    sahm_threshold: float = 1.0


@dataclass
class V2Config(Config):
    split: int = 10
    buy_rsi: int = 30
    sell_rsi: int = 80

File: src/configs.cpp
#include "configs.h"
#include <sstream>
#include <iomanip>
#include <functional>

Config Config::from_map(const std::unordered_map<std::string, double>& source) {
    Config config;
    
    auto it = source.find("term");
    if (it != source.end()) config.term = static_cast<int>(it->second);
    
    it = source.find("margin");
    if (it != source.end()) config.margin = it->second;
    
    it = source.find("bullish_rsi");
    if (it != source.end()) config.bullish_rsi = static_cast<int>(it->second);
    
    it = source.find("burst_urate");
    if (it != source.end()) config.burst_urate = it->second;
    
    it = source.find("burst_scale");
    if (it != source.end()) config.burst_scale = it->second;
    
    it = source.find("burst_vol");
    if (it != source.end()) config.burst_vol = static_cast<int>(it->second);
    
    it = source.find("sell_base");
    if (it != source.end()) config.sell_base = it->second;
    
    it = source.find("sell_limit");
    if (it != source.end()) config.sell_limit = it->second;
    
    it = source.find("sahm_threshold");
    if (it != source.end()) config.sahm_threshold = it->second;
    
    it = source.find("rsi_threshold");
    if (it != source.end()) config.rsi_threshold = it->second;
    
    it = source.find("buy_splits");
    if (it != source.end()) config.buy_splits = static_cast<int>(it->second);

    
    return config;
}

std::size_t Config::hash() const {
    std::size_t h = 0;
    h = std::hash<int>{}(term) ^ (h << 1);
    h = std::hash<double>{}(margin) ^ (h << 1);
    h = std::hash<int>{}(bullish_rsi) ^ (h << 1);
    h = std::hash<double>{}(burst_urate) ^ (h << 1);
    h = std::hash<double>{}(burst_scale) ^ (h << 1);
    h = std::hash<int>{}(burst_vol) ^ (h << 1);
    h = std::hash<double>{}(sell_base) ^ (h << 1);
    h = std::hash<double>{}(sell_limit) ^ (h << 1);
    h = std::hash<double>{}(sahm_threshold) ^ (h << 1);
    h = std::hash<double>{}(rsi_threshold) ^ (h << 1);
    h = std::hash<int>{}(buy_splits) ^ (h << 1);
    return h;
}

std::string Config::to_string() const {
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(2);
    
    oss << "term: " << term << ", ";
    oss << "margin: " << margin << ", ";
    oss << "bullish_rsi: " << bullish_rsi << ", ";
    oss << "burst_urate: " << burst_urate << ", ";
    oss << "burst_scale: " << burst_scale << ", ";
    oss << "burst_vol: " << burst_vol << ", ";
    oss << "sell_base: " << sell_base << ", ";
    oss << "sell_limit: " << sell_limit << ", ";
    oss << "sahm_threshold: " << sahm_threshold << ", ";
    oss << "rsi_threshold: " << rsi_threshold << ", ";
    oss << "buy_splits: " << buy_splits;
    
    return oss.str();
}

bool Config::operator==(const Config& other) const {
    return term == other.term &&
           margin == other.margin &&
           bullish_rsi == other.bullish_rsi &&
           burst_urate == other.burst_urate &&
           burst_scale == other.burst_scale &&
           burst_vol == other.burst_vol &&
           sell_base == other.sell_base &&
           sell_limit == other.sell_limit &&
           sahm_threshold == other.sahm_threshold &&
           rsi_threshold == other.rsi_threshold &&
           buy_splits == other.buy_splits;
}